src/liballoc/string.rs

```diff
@@ -1225,6 +1225,62 @@ impl String {
         ch
     }
 
+    /// Remove all matches of pattern `pat` in the `String`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(string_remove_matches)]
+    /// let mut s = String::from("Trees are not green, the sky is not blue.");
+    /// s.remove_matches("not ");
+    /// assert_eq!("Trees are green, the sky is blue.", s);
+    /// ```
+    ///
+    /// Matches will be detected and removed iteratively, so in cases where
+    /// patterns overlap, only the first pattern will be removed:
+    ///
+    /// ```
+    /// #![feature(string_remove_matches)]
+    /// let mut s = String::from("banana");
+    /// s.remove_matches("ana");
+    /// assert_eq!("bna", s);
+    /// ```
+    #[unstable(feature = "string_remove_matches", reason = "new API", issue = "none")]
+    pub fn remove_matches<'a, P>(&'a mut self, pat: P)
+    where
+        P: for<'x> Pattern<'x>,
+    {
+        use core::str::pattern::Searcher;
+
+        let matches = {
+            let mut searcher = pat.into_searcher(self);
+            let mut matches = Vec::new();
```

Comment:
... I don't think using a Vec here counts as "does not use any allocation" (the original promise in #50206).

Reply:
Originally I couldn't think of a way to do this without allocating a Vec to hold the match start/ends, but I think I have an idea now that may work. I'll try it when I get off work today in about 6 hours.

Reply:
I went ahead and tested my idea and unfortunately it doesn't work. The way I can see this you either have to:
a. Store the match start/end pairs somewhere, then start modifying the string, as I've done here
or
b. Modify the string during searching.

The problem with b is that the searcher will end up missing matches as parts it would've matched on are potentially moved behind where it's looking.

Here's the code I tried, for reference:
```rust
use core::str::pattern::Searcher;
let vec_ptr = self.vec.as_mut_ptr();

let len = self.len();
let mut shrunk_by = 0;
let mut searcher = pat.into_searcher(self);

// SAFETY: start and end will be on utf8 byte boundaries per
// the Searcher docs
unsafe {
    while let Some((start, end)) = searcher.next_match() {
        ptr::copy(
            vec_ptr.add(end - shrunk_by),
            vec_ptr.add(start - shrunk_by),
            len - end,
        );
        shrunk_by += end - start;
    }
    drop(searcher);
    self.vec.set_len(len - shrunk_by);
}
```

Reply:
i see value in this new method, even if it does allocate. i'm not on the libs team so don't weight my opinion too heavily, but my vote would be to merge a working implementation, and then open up a follow up issue for removing the allocation

Reply:
I do think this method is better, I don't see why do we need another `Vec` for this. The only good reason I see doing a two-pass method like here is used in SIMD JSON which the first pass identify the points which makes deserialization faster.